flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPoolTest.java

```diff
@@ -518,4 +528,176 @@ public void go() throws Exception {
+		final int numBuffers = 10;
+		final int numberOfSegmentsToRequest = 5;
+		final Duration requestSegmentsTimeout = Duration.ofSeconds(10L);
+
+		final NetworkBufferPool globalPool = new NetworkBufferPool(
+				numBuffers,
+				128,
+				numberOfSegmentsToRequest,
+				requestSegmentsTimeout);
+
+		try {
+			assertTrue(globalPool.isAvailable().isDone());
+
+			List<MemorySegment> segments = new ArrayList<>(numberOfSegmentsToRequest);
+			for (int i = 0; i < numberOfSegmentsToRequest; ++i) {
+				MemorySegment segment = globalPool.requestMemorySegment();
+				assertNotNull(segment);
+				segments.add(segment);
+				assertTrue(globalPool.isAvailable().isDone());
+			}
+
+			final List<MemorySegment> exclusiveSegments = globalPool.requestMemorySegments();
+			assertEquals(numberOfSegmentsToRequest, exclusiveSegments.size());
+
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(0, globalPool.getNumberOfAvailableMemorySegments());
+
+			final AtomicBoolean completeFlag = new AtomicBoolean(false);
+			CompletableFuture<?> availableFuture = globalPool.isAvailable();
+			availableFuture.whenComplete((ignored, throwable) -> completeFlag.set(true));
+
+			// recycle one buffer
+			globalPool.recycle(segments.get(0));
+			assertTrue(completeFlag.get());
+			assertTrue(availableFuture.isDone());
+			assertTrue(globalPool.isAvailable().isDone());
+			assertEquals(1, globalPool.getNumberOfAvailableMemorySegments());
+
+			CheckedThread asyncRequest = new CheckedThread() {
+				@Override
+				public void go() throws Exception {
+					exclusiveSegments.addAll(globalPool.requestMemorySegments());
+				}
+			};
+			asyncRequest.start();
+
+			// wait until no buffer is available
+			final Deadline deadline = Deadline.fromNow(Duration.ofSeconds(10L));
+			while (globalPool.getNumberOfAvailableMemorySegments() > 0) {
+				Thread.sleep(50);
+				if (!deadline.hasTimeLeft()) {
+					fail("Waiting timeout.");
+				}
+			}
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(numberOfSegmentsToRequest, exclusiveSegments.size());
+
+			for (int i = 1; i < numberOfSegmentsToRequest; ++i) {
+				globalPool.recycle(segments.get(i));
+			}
+			segments.clear();
+
+			asyncRequest.sync();
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(numBuffers, exclusiveSegments.size());
+			assertFalse(globalPool.isAvailable().isDone());
+
+			globalPool.recycleMemorySegments(exclusiveSegments);
+			exclusiveSegments.clear();
+			assertTrue(globalPool.isAvailable().isDone());
+			assertEquals(numBuffers, globalPool.getNumberOfAvailableMemorySegments());
+
+		} finally {
+			globalPool.destroy();
+		}
+	}
+
+	/**
+	 * Tests that blocking request of multi local buffer pools can be fulfilled by recycled segments
+	 * to the global network buffer pool.
+	 */
+	@Test
+	public void testBlockingRequestFromMultiLocalBufferPool() throws Exception {
+		final int localBufferPoolSize = 10;
+		final int numLocalBufferPool = 10;
```

Comment:
we can reduce the `numLocalBufferPool` as 2 instead to achieve the purpose to avoid creating 10 threads in `ExecutorService`.